File: js/test/pipe/pipe.test.js
const { expect } = require('chai');
const _ = require('lodash');
const sleep = require('../../lib/sleep');

const pipe = require('../../pipe');
const randomString = require('../helpers/randomString');

const msgcount = 10;
const pipecount = 3;

[
  pipe.NewGRPCPipe,
  pipe.NewHTTPPipe
].forEach((Pipe, ndx) => {
  describe('Pipe - ' + ndx, () => {
    describe('constructor', () => {
      it('should blow up if no apikey', async () => {
        try {
          Pipe();
          expect(1).eq(0);
        } catch (er) {
          expect(1).eq(1);
        };
      });
    });

    describe('basic interactions', () => {
      let pipe;
      let pipename = randomString(8);
      beforeEach(async () => {
        pipe = Pipe(pipename, process.env.PIPELINR_API_KEY);
      });

      describe('receiveOptions', () => {
        it('shoud set receive opts', async () => {
          expect(pipe.receiveOptions.autoAck).eq(false);
          expect(pipe.receiveOptions.block).eq(false);
          expect(pipe.receiveOptions.pipe).eq(pipename);
          expect(pipe.receiveOptions.count).eq(1);
          expect(pipe.receiveOptions.timeout).eq(0);
          expect(pipe.receiveOptions.redeliveryTimeout).eq(0);

          pipe.receiveOptions = {autoAck: true, count: 5};

          expect(pipe.receiveOptions.autoAck).eq(true);
          expect(pipe.receiveOptions.block).eq(false);
          expect(pipe.receiveOptions.pipe).eq(pipename);
          expect(pipe.receiveOptions.count).eq(5);
          expect(pipe.receiveOptions.timeout).eq(0);
          expect(pipe.receiveOptions.redeliveryTimeout).eq(0);  
        });
      });

      describe('send/recv', () => {
        beforeEach(async () => {
          pipe.receiveOptions = {
            autoAck: true,
            block: true,
            count: 1,
            timeout: 10,
            redeliveryTimeout: 50,
          };
        });

        it('should send one, then receive, ack, log, and complete one', async () => {
          /*  send  */
          const id = await pipe.send({
            Payload: JSON.stringify({foo: 'bar'}),
            Route: [pipename, 'some', 'route'],
          });

          expect(id).not.eq('');
          expect(typeof id).eq('string');

          /*  recv  */
          const elms = await pipe.fetch();
          expect(elms.length).eq(1);

          const elm = elms[0];
          expect(elm.id).eq(id);
          expect(elm.message.route).deep.eq([pipename, 'some', 'route']);
          expect(elm.message.payload.foo).eq('bar');
          expect(elm.message.decoratedPayload.foo).eq('bar');

          /*  ack  */
          const ackres = await pipe.ack(elm.id);
          expect(ackres).eq(true);

          /*  log  */
          const logres = await pipe.log(elm.id, 3, `some message from ${pipename}`);
          expect(logres).eq(true);
          
          /*  addSteps  */
          const stepsres = await pipe.addSteps(id, ['flip', 'flap']);
          expect(stepsres).eq(true);

          /*  decorate  */
          const decres = await pipe.decorate(id, [
            {Key: 'somekey', Value: 'somevalue'},
            {Key: 'otherkey', Value: 'otherval'}
          ]);
          expect(decres).not.eq(undefined);
          expect(decres).not.eq(null);
          expect(decres.length).eq(2);
          expect(decres[0]).eq(true);
          expect(decres[1]).eq(true);

          /*  complete  */
          const completeres = await pipe.complete(id);
          expect(completeres).eq(true);
        });

        it('should omit fields when configured to', async () => {
          pipe.receiveOptions = {
            excludeRouting: true,
            excludeRouteLog: true,
            excludeDecoratedPayload: true,
          }

          const id = await pipe.send({
            Payload: JSON.stringify({foo: 'bar'}),
            Route: [pipename, 'some', 'route'],
          });

          expect(id).not.eq('');
          expect(typeof id).eq('string');

          expect(await pipe.decorate(id, [
            {Key: 'somekey', Value: 'somevalue'},
            {Key: 'otherkey', Value: 'otherval'}
          ])).deep.eq([true, true]);

          const elms = await pipe.fetch();
          expect(elms.length).eq(1);
          const elm = elms[0];
          expect(elm.id).eq(id);
          expect(elm.message.route).deep.eq([])
          expect(elm.message.decoratedPayload).deep.eq({});
          expect(elm.message.routeLog).deep.eq([]);
        });
      });
    });

    describe('flow', () => {
      let pipes = [];
      let pipenamebase = `${randomString(8)}`;
      beforeEach(async () => {
        pipes.push(Pipe(pipenamebase + '-step-1', process.env.PIPELINR_API_KEY));
        pipes.push(Pipe(pipenamebase + '-step-2', process.env.PIPELINR_API_KEY));
        pipes.push(Pipe(pipenamebase + '-step-3', process.env.PIPELINR_API_KEY));

        pipes.forEach((p) => {
          p.receiveOptions = {
            autoAck: false,
            block: false,
            count: 10,
            timeout: 10,
            redeliveryTimeout: 60,
          };

          p.start();
        });
      });

      afterEach(async () => {
        pipes.forEach((p) => {
          p.stop();
        });
      });

      it('should update the message as it goes along', async () => {
        const id = await pipes[0].send({
          Payload: JSON.stringify({foo: 'blip'}),
          Route: [
            pipenamebase + '-step-1',
            pipenamebase + '-step-2',
            pipenamebase + '-step-3',
        ]});

        const stage1 = await pipes[0].next();
        expect(stage1.message.decoratedPayload).deep.eq(stage1.message.payload);
        expect((await pipes[0].ack(stage1.id))).eq(true);
        expect((
          await pipes[0].decorate(
            stage1.id, [{Key: 'stage1', Value: 'yep'}])
            ).length
            ).eq(1);
        expect((await pipes[0].complete(stage1.id))).eq(true);

        const stage2 = await pipes[1].next();
        expect(stage2.message.decoratedPayload.stage1).eq('yep');
        expect((await pipes[1].decorate(stage2.id, [{Key: 'stage1', Value: 'overwritten'}, {Key: 'stage2', Value: 'done'}])).length).eq(2);
        expect((await pipes[1].complete(stage2.id))).eq(true);
        
        const stage3 = await pipes[2].next()
        expect(stage3.message.decoratedPayload.stage1).eq('overwritten');
        expect(stage3.message.decoratedPayload.stage2).eq('done');
        expect((await pipes[2].complete(stage3.id))).eq(true);
      });
    });

    describe('large number of messages and pipes', () => {
      let pipes = [];
      let pipenamebase = `${randomString(8)}`;
      
      let totalShouldProcess = msgcount * pipecount;

      beforeEach(async () => {
        for (let i = 0; i < pipecount; i++) {
          const p = Pipe(`${pipenamebase}-bigstep-${i}`, process.env.PIPELINR_API_KEY)
          p.receiveOptions = {
            autoAck: false,
            block: false,
            count: 50,
            timeout: 1,
            redeliveryTimeout: 10,
          };

          p.start(msgcount);

          pipes.push(p);
        }
      });

      afterEach(async () => {
        pipes.forEach((p) => {
          p.stop();
        });
      });

      it('should process through everything', async () => {
        let totalProcessed = 0;
        let events = {};

        const pushem = async () => {
          const route = pipes.map(p => p.name);

          const topush = [];

          for (let i = 0; i < msgcount; i++) {
            topush.push({
              Payload: JSON.stringify({messagenum: i}),
              Route: route,
            });
            // await pipes[0].send();
          }

          console.log('gonna push some messages', topush.length);

          const work = async () => {
            let pushedcount = 0;
            const st = new Date();
            while(topush.length > 0) {
              const msg = topush.shift();
              await pipes[0].send(msg);
              pushedcount += 1;
            }
            console.log('worker done, messages pushed by this worker', pushedcount, 'in', new Date() - st, 'ms');
          }

          for(let i = 0; i < 10; i++) {
            work();
          }
        }

        pushem();


        while(totalProcessed < totalShouldProcess) {
          // console.log('processed', totalProcessed, 'of', totalShouldProcess, totalProcessed/totalShouldProcess * 100.0, '%');
          for(let pipndx = 0; pipndx < pipes.length; pipndx++) {
            const pipe = pipes[pipndx];
            const thendx = pipndx;
            pipe.next().then((elm) => {
              return pipe.decorate(elm.id, [{Key: `step-${thendx}`, Value: `"${thendx}"`}]).then((res) => {
                return pipe.complete(elm.id).then((res) => {
                  events[elm.id] = elm;
                  totalProcessed++;
                });
              })
            });
          }
          // console.log('sleeping', Object.keys(events).length)
          await sleep(100); 
        }
        
        _.forEach(events, (event, id) => {
          expect(event.message.completedSteps.length).eq(pipes.length-1);
          for(let pipndx = 0; pipndx < pipes.length - 1; pipndx++) {
            expect(event.message.decoratedPayload[`step-${pipndx}`]).eq(`${pipndx}`); 
          }         
        })
      })
    });

    describe('decoration interactions', () => {
      let pipe;
      beforeEach(async () => {
        pipe = Pipe('decoration-interactions', process.env.PIPELINR_API_KEY)
      });

      it('should decorate with keys and subkeys, and expose a way to get specific decorations', async () => {
        const mid = await pipe.send({
          Payload: JSON.stringify({foo: 'blap'}),
          Route: ['decoration-interactions'],
        });

        expect(await pipe.decorate(mid, [
          {Key: 'numeric', Value: `1`},
          {Key: 'flap', Value: 'bar'}, 
          {Key: 'nested.key', Value: 'fff'}, 
          {Key: 'deeply.nested.key', Value: 'fsdf'},
          {Key: 'object', Value: JSON.stringify({
            deep: {
              fields:{
                are: {
                  ok:{
                    with: 'me'
                  }
                }
              }
            }
          })}
        ])).deep.eq([true, true, true, true, true]);

        const decs = await pipe.getDecorations(mid, ['numeric', 'nested.key', 'object']);
        expect(decs.numeric).eq(1);
        expect(decs.nested.key).eq('fff');
        expect(decs.object.deep.fields.are.ok.with).eq('me');
      });
    });
  });
});

File: go/pipe/pipe_test.go
package pipe

import (
	"fmt"
	"sync"
	"testing"
	"time"

	"github.com/nochte/pipelinr-clients/go/lib"
	"github.com/nochte/pipelinr-clients/go/pipe/drivers"
	"github.com/nochte/pipelinr-protocol/protobuf/messages"
	"github.com/nochte/pipelinr-protocol/protobuf/pipes"
	. "github.com/smartystreets/goconvey/convey"
	"github.com/tidwall/gjson"
)

func TestPipe(t *testing.T) {
	msgcount := 11
	pipecount := 4

	Convey("Pipe tests", t, func() {
		type TestDefinition struct {
			getdriver func() drivers.Driver
			name      string
		}
		for _, td := range []TestDefinition{
			{name: "grpc pipe", getdriver: func() drivers.Driver {
				return drivers.NewGRPCDriver("", "")
			}},
			{name: "http driver", getdriver: func() drivers.Driver {
				return drivers.NewHTTPDriver("", "")
			}},
		} {
			Convey(td.name, func() {
				driver := td.getdriver()

				Convey("basic interactions", func() {
					pipename := lib.GenerateRandomString(8)
					pipe := New(driver, pipename)

					Convey("receiveOptions", func() {
						So(pipe.ReceiveOptions().GetAutoAck(), ShouldEqual, false)
						So(pipe.ReceiveOptions().GetBlock(), ShouldEqual, false)
						So(pipe.ReceiveOptions().GetPipe(), ShouldEqual, pipename)
						So(pipe.ReceiveOptions().GetCount(), ShouldEqual, 1)
						So(pipe.ReceiveOptions().GetTimeout(), ShouldEqual, 0)
						So(pipe.ReceiveOptions().GetRedeliveryTimeout(), ShouldEqual, 0)

						tr := true
						ct := int32(5)
						pipe.SetReceiveOptions(&ct, nil, nil, &tr, nil, nil, nil, nil)

						So(pipe.ReceiveOptions().GetAutoAck(), ShouldEqual, true)
						So(pipe.ReceiveOptions().GetBlock(), ShouldEqual, false)
						So(pipe.ReceiveOptions().GetPipe(), ShouldEqual, pipename)
						So(pipe.ReceiveOptions().GetCount(), ShouldEqual, 5)
						So(pipe.ReceiveOptions().GetTimeout(), ShouldEqual, 0)
						So(pipe.ReceiveOptions().GetRedeliveryTimeout(), ShouldEqual, 0)
					})
					Convey("send/recv", func() {
						tru := true
						one := int32(1)
						ten := int64(10)
						fifty := int64(50)
						pipe.SetReceiveOptions(&one, &ten, &fifty, &tru, &tru, nil, nil, nil)

						id, er := pipe.Send(`{"foo":"bar"}`, []string{pipename, "some", "route"})
						So(er, ShouldBeNil)
						So(len(id), ShouldBeGreaterThan, 1)

						// recv
						elms, er := pipe.Fetch()
						So(er, ShouldBeNil)
						So(len(elms), ShouldEqual, 1)
						elm := elms[0]
						So(elm.GetStringId(), ShouldEqual, id)
						So(elm.GetMessage().GetRoute(), ShouldResemble, []string{pipename, "some", "route"})
						So(gjson.Get(elm.GetMessage().GetPayload(), "foo").String(), ShouldEqual, "bar")
						So(gjson.Get(elm.GetMessage().GetDecoratedPayload(), "foo").String(), ShouldEqual, "bar")

						// ack
						So(pipe.Ack(id), ShouldBeNil)

						// log
						So(pipe.Log(id, 3, fmt.Sprintf("some message from %v", pipename)), ShouldBeNil)

						// addsteps
						So(pipe.AddSteps(id, []string{"flip", "flap"}), ShouldBeNil)

						// decorate
						res := pipe.Decorate(id, []*pipes.Decoration{
							{Key: "somekey", Value: "somevalue"},
							{Key: "otherkey", Value: "otherval"}})

						So(len(res), ShouldEqual, 2)
						So(res[0], ShouldBeNil)
						So(res[1], ShouldBeNil)

						// complete
						So(pipe.Complete(id), ShouldBeNil)
					})
				})
				Convey("flow", func() {
					pipenamebase := lib.GenerateRandomString(8)
					ppipes := []*Pipe{
						New(driver, fmt.Sprintf("%v-step-1", pipenamebase)),
						New(driver, fmt.Sprintf("%v-step-2", pipenamebase)),
						New(driver, fmt.Sprintf("%v-step-3", pipenamebase)),
					}
					for _, p := range ppipes {
						i32ten := int32(10)
						i64ten := int64(10)
						i64sixty := int64(60)
						p.SetReceiveOptions(&i32ten, &i64ten, &i64sixty, nil, nil, nil, nil, nil)

						go p.Start(0)
					}

					Reset(func() {
						for _, p := range ppipes {
							p.Stop()
						}
					})

					_, er := ppipes[0].Send(`{"foo":"blip"}`, []string{
						fmt.Sprintf("%v-step-1", pipenamebase),
						fmt.Sprintf("%v-step-2", pipenamebase),
						fmt.Sprintf("%v-step-3", pipenamebase)})
					So(er, ShouldBeNil)

					stage1 := <-ppipes[0].Chan()
					So(stage1.GetMessage().GetDecoratedPayload(), ShouldEqual, `{"foo":"blip"}`)
					So(ppipes[0].Ack(stage1.GetStringId()), ShouldBeNil)

					decres := ppipes[0].Decorate(stage1.GetStringId(), []*pipes.Decoration{
						{Key: "stage1", Value: "yep"}})
					So(decres[0], ShouldBeNil)

					So(ppipes[0].Complete(stage1.GetStringId()), ShouldBeNil)

					stage2 := <-ppipes[1].Chan()
					So(gjson.Get(stage2.GetMessage().GetDecoratedPayload(), "stage1").String(), ShouldEqual, "yep")
					So(ppipes[1].Decorate(stage2.GetStringId(), []*pipes.Decoration{
						{Key: "stage1", Value: "overwritten"},
						{Key: "stage2", Value: "done"}}), ShouldResemble, []error{nil, nil})
					So(ppipes[1].Complete(stage2.GetStringId()), ShouldBeNil)

					stage3 := <-ppipes[2].Chan()
					So(gjson.Get(stage3.GetMessage().GetDecoratedPayload(), "stage1").String(), ShouldEqual, "overwritten")
					So(gjson.Get(stage3.GetMessage().GetDecoratedPayload(), "stage2").String(), ShouldEqual, "done")
					So(ppipes[2].Complete(stage3.GetStringId()), ShouldBeNil)
				})
				Convey("large number of messages and pipes", func() {
					pipenamebase := lib.GenerateRandomString(8)
					// totalShouldProcess := msgcount * pipecount

					// setup pipes
					ppipes := make([]*Pipe, 0, pipecount)
					i32fifty := int32(50)
					i64ten := int64(10)
					for i := 0; i < pipecount; i++ {
						p := New(driver, fmt.Sprintf("%v-bigstep-%v", pipenamebase, i))
						p.SetReceiveOptions(&i32fifty, nil, &i64ten, nil, nil, nil, nil, nil)
						go p.Start(0)

						ppipes = append(ppipes, p)
					}

					topush := make(chan *messages.MessageEnvelop, msgcount)
					pushwg := sync.WaitGroup{}
					for i := 0; i < 1; i++ {
						pushwg.Add(1)

						go func(ndx int) {
							defer pushwg.Done()
							startTime := time.Now()
							pushcount := 0
							for msg := range topush {
								ppipes[0].Send(msg.GetPayload(), msg.GetRoute())
								pushcount++
							}
							fmt.Println(ndx, "worker done - messages pushed by this worker", pushcount, "in", time.Since(startTime))
						}(i)
					}

					route := make([]string, 0, len(ppipes))
					for _, p := range ppipes {
						route = append(route, p.Name())
					}

					for i := 0; i < msgcount; i++ {
						topush <- &messages.MessageEnvelop{
							Route:   route,
							Payload: fmt.Sprintf(`{"messagenum":"%v"}`, i)}
					}
					close(topush)

					// Println("messages enqueued, waiting til done")
					// pushwg.Wait()
					// Println("done pushing")

					totalShouldProcess := pipecount * msgcount

					elmchan := make(chan *messages.Event, totalShouldProcess)

					pwg := sync.WaitGroup{}
					for pndx := range ppipes {
						pwg.Add(1)
						go func(ppipe *Pipe, ndx int) {
							defer pwg.Done()
							ch := ppipe.Chan()
							for elm := range ch {
								// fmt.Println(ppipe.Name(), "got a message", elm.Message.Payload)
								ers := ppipe.Decorate(elm.GetStringId(), []*pipes.Decoration{
									{Key: fmt.Sprintf("step-%v", ndx), Value: fmt.Sprintf(`"%v"`, ndx)}})
								for _, er := range ers {
									if er != nil {
										fmt.Println(ppipe.Name(), "error decorating", er.Error())
									}
								}
								if er := ppipe.Complete(elm.GetStringId()); er != nil {
									fmt.Println(ppipe.Name(), "error completing", er.Error())
								}
								elmchan <- elm
							}
						}(ppipes[pndx], pndx)
					}

					events := make(map[string]*messages.Event)
					totalProcessed := 0
					go func() {
						for elm := range elmchan {
							events[elm.GetStringId()] = elm
							totalProcessed++
						}
					}()

					for totalProcessed < totalShouldProcess {
						time.Sleep(time.Millisecond * 250)
					}
					for pndx := range ppipes {
						ppipes[pndx].Stop()
					}

					Println("all done setting up, waiting for processing to complete")
					pwg.Wait()
					close(elmchan)

					for _, evt := range events {
						So(len(evt.GetMessage().GetCompletedSteps()), ShouldEqual, len(ppipes)-1)

						for ndx := 0; ndx < len(ppipes)-1; ndx++ {
							So(gjson.Get(evt.GetMessage().GetDecoratedPayload(), fmt.Sprintf("step-%v", ndx)).String(), ShouldEqual, fmt.Sprintf(`%v`, ndx))
						}
					}
				})
			})
		}
	})
}
